### PR TITLE
Support for Node 18

### PR DESCRIPTION
**Pull request by feuGeneA**

### CODE REVIEW

File: .github/workflows/e2e-ci.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node: [12, 14, 16]
+        node: [14, 16, 18]
```
Comment (alcuadrado) [inline .github/workflows/e2e-ci.yml:34]: At first, I was "wait, we still support node 12!", but we are dropping its support in 13 days, so yeah, let's do this 🥳

File: docs/src/content/config/index.md [hunk 2]
```diff
@@ -119,7 +119,7 @@ module.exports = {
 
 ```js
 {
-    localhost: {
+    127.0.0.1: {
```
Comment (alcuadrado) [inline docs/src/content/config/index.md:122]:
```suggestion
    localhost: {
```

This is the name of the network.

File: packages/hardhat-core/CHANGELOG.md [hunk 3]
```diff
@@ -146,7 +146,7 @@
 - 8076c43b: Fixed how the cummulative gas is computed for receipts and added a missing field (Thanks @ngotchac!)
 - e6362902: Display similar artifact names in error output if given name is not found (#201)
 - e087bd0b: Improve validation of private keys in the Hardhat config
-- aa1a0080: Fix an issue with new versions of Node.js that prevented clients from connecting to Hardhat's node using 127.0.0.1
+- aa1a0080: Fix an issue with new versions of Node.js that prevented clients from connecting to Hardhat's node using localhost
```
Comment (alcuadrado) [inline packages/hardhat-core/CHANGELOG.md:149]:
```suggestion
- aa1a0080: Fix an issue with new versions of Node.js that prevented clients from connecting to Hardhat's node using 127.0.0.1
```

Let's not change the changelogs, as they are useful as a record of what happened.

File: packages/hardhat-core/package.json [hunk 4]
```diff
@@ -23,7 +23,7 @@
     "hardhat": "internal/cli/cli.js"
   },
   "engines": {
-    "node": "^12.0.0 || ^14.0.0 || ^16.0.0"
+    "node": "^14.0.0 || ^16.0.0 || ^18.0.0"
```
Comment (alcuadrado) [inline packages/hardhat-core/package.json:26]:
```suggestion
    "node": "^12.0.0 || ^14.0.0 || ^16.0.0 || ^18.0.0"
```

This is unfortunate, but let's play nicely here.

File: packages/e2e/package.json [hunk 5]
```diff
@@ -32,7 +32,7 @@
     "eslint-plugin-import": "2.24.1",
     "eslint-plugin-prettier": "3.4.0",
     "fs-extra": "^7.0.1",
-    "mocha": "^9.2.0",
+    "mocha": "^10.0.0",
```
Comment (alcuadrado) [inline packages/e2e/package.json:35]: Was this change required? Why?

File: packages/hardhat-ethers/test/ethers-provider-wrapper.ts [hunk 6]
```diff
@@ -12,7 +12,9 @@ describe("Ethers provider wrapper", function () {
   useEnvironment("hardhat-project");
 
   beforeEach(function () {
-    realProvider = new ethers.providers.JsonRpcProvider();
+    realProvider = new ethers.providers.JsonRpcProvider(
+      "http://127.0.0.1:8545"
```
Comment (alcuadrado) [inline packages/hardhat-ethers/test/ethers-provider-wrapper.ts:16]: Good catch! 🙌

File: packages/hardhat-ganache/test/index.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@ import { GanacheService } from "../src/ganache-service";
 
 import { useEnvironment } from "./helpers";
 
-describe("Ganache plugin with empty configs", function () {
+describe.skip("Ganache plugin with empty configs", function () {
```
Comment (alcuadrado) [inline packages/hardhat-ganache/test/index.ts:7]:
```suggestion
// Note: Skipping as this plugin hasn't been updated to the latest Ganache version,
//   so it doesn't work well with Node 18
describe.skip("Ganache plugin with empty configs", function () {
```

File: packages/hardhat-ganache/test/index.ts [hunk 8]
```diff
@@ -49,7 +49,7 @@ describe("Ganache plugin with empty configs", function () {
   });
 });
 
-describe("Ganache plugin with custom configs", function () {
+describe.skip("Ganache plugin with custom configs", function () {
```
Comment (alcuadrado) [inline packages/hardhat-ganache/test/index.ts:52]:
```suggestion
// Note: Skipping as this plugin hasn't been updated to the latest Ganache version,
//   so it doesn't work well with Node 18
describe.skip("Ganache plugin with custom configs", function () {
```

File: packages/e2e/package.json (hunk 5 above)
Comment (alcuadrado) [reply packages/e2e/package.json:35]: I'm actually ok with it, but I want to understand how this may affect existing users.

**Comment by alcuadrado** [comment]
Great work!

I left a few minor comments.

We need to add changesets to many packages. I can do it.

File: packages/hardhat-ganache/src/ganache-service.ts [hunk 9]
```diff
@@ -202,7 +202,7 @@ export class GanacheService {
 
     // Validate and parse hostname and port from URL (this validation is priority)
     const url = new URL(options.url);
-    if (url.hostname !== "localhost" && url.hostname !== "127.0.0.1") {
+    if (url.hostname !== "127.0.0.1") {
```
Comment (alcuadrado) [inline packages/hardhat-ganache/src/ganache-service.ts:205]:
```suggestion
    if (url.hostname !== "localhost" && url.hostname !== "127.0.0.1") {
```

I think we can revert this and not publish a new version, as it's not very interesting.

File: packages/e2e/package.json (hunk 5 above)
Comment (morgansliman) [reply packages/e2e/package.json:35]: I'm not necessarily sure if it was "required" per se, but mocha 10 is basically their version of the same upgrade we're doing here (i.e. deprecating node 12 support and adding node 18 support to their CI), so I felt like it made sense to include it.

Comment (alcuadrado) [reply packages/e2e/package.json:35]: Makes sense

**Review by alcuadrado — APPROVED** [review]
